### ISSUE TITLE
Invalid TypeDoesNotContainType error

### ISSUE THREAD

**Issue by azjezz** [vimeo/psalm#2662]
code : 
```php
final class User {
    private string $username;
    
    public function setUsername(string $username): void {
        $this->username = $username;
    }
    
    public function getUsername(): ?string
    {
        return $this->username ?? null;
    }
}

$user = new User();

echo $user->getUsername() ?? 'RETURNED NULL ( uninitialized )'.PHP_EOL;
$user->setUsername('foo');
echo $user->getUsername();
```

using `??` is valid to check if a property has been initialized in php, since `??` checks if the variable exist ( has been initialized ) and whether it is null, so using `??` on an uninitialized property is *valid*, and psalm should not complain about uninitialized property in this case either.

see : https://psalm.dev/r/56548f7f4f
see : https://3v4l.org/mFXLR

**Comment by muglug** [comment]
Maybe this seems overly prescriptivist of me, but this is _really_ not how you should be using properties - instead make the property type nullable, and the code is easier to understand: https://psalm.dev/r/98cd7901d7

**Comment by azjezz** [comment]
@muglug yes, but this means that to property can be "uninitialized" at any point later, by setting its value to null. having `string` instead of `?string` forbids this, even with reflections : 

example : 
```php
<?php

final class Foo {
  private ?string $foo = null;
  
  public function setFoo(string $foo): void {
      $this->foo = $foo;
  }
  
  public function getFoo(): ?string {
      return $this->foo ?? null;
  }
}

final class Bar {
  private string $bar;
  
  public function setBar(string $bar): void {
      $this->bar = $bar;
  }
  
  public function getBar(): ?string {
      return $this->bar ?? null;
  }
}

$bar = new Bar();
$bar->setBar('hello');

$foo = new Foo();
$foo->setFoo('hello');

$reflection = new \ReflectionObject($foo);
$property = $reflection->getProperty('foo');
$property->setAccessible(true);
$property->setValue($foo, null);

\var_dump($property->getValue($foo));

$reflection = new \ReflectionObject($bar);
$property = $reflection->getProperty('bar');
$property->setAccessible(true);
$property->setValue($bar, null); // throws !
```

3v4l : https://3v4l.org/iSXdn

**Comment by azjezz** [comment]
maybe a new annotation can tell the difference here ? similar to hacks `<<__LateInit>>` attribute ?

see : https://docs.hhvm.com/hack/attributes/predefined-attributes#__lateinit

**Comment by muglug** [comment]
You can still always `unset` a property, so it’s not even safer.

Adding a `<<__LateInit>>` annotation in Hack is largely equivalent to `@psalm-suppress MissingConstructor`, I think.

**Comment by azjezz** [comment]
> You can still always unset a property, so it’s not even safer.

note that i said `"uninitialized"` not actually uninitialized, because setting its value to `null` doesn't mean that the property has been uninitialized from the engines point ( `$propertyReflection->isInitialized($foo)` would return `true` ) 

meanwhile `unset()` actually un-initializes the property ( `$propertyReflection->isInitialized($foo)` would return `false` ) 

> Adding a <<__LateInit>> annotation in Hack is largely equivalent to @psalm-suppress MissingConstructor, I think.

and allowing the use of `??` and `isset` when the type is not nullable maybe ? :) 

this is not possible in hack btw, because hhvm doesn't check if property has been initialized or not with `??`, actually,the only way to do so now in hack is using `isset`.

```hack
final class User {
  <<__LateInit>>
  private string $username;

  public function setUsername(string $username): void {
     $this->username = $username;
  }

  public function getUsername(): ?string {
    return isset($this->username) : $this->username : null;
  }
}

<<__EntryPoint>>
function main(): void {
  $user = new User();
  \var_dump($user->getUsername());
  $user->setUsername('foo');
  \var_dump($user->getUsername());
}
```

in this case, the type checker only complains about using `isset`, but at runtime, it is the same behaviour as `$this->username ?? null` in php.

```
➜  hack-psalm hh_client nullable-late-init.hack
Typing[4016] isset tends to hide errors due to variable typos and so is limited to dynamic checks in strict mode
   --> nullable-late-init.hack
 10 |     return isset($this->username) ? $this->username : null;
    |            ^^^^^^^^^^^^^^^^^^^^^^

1 error found.
➜  hack-psalm 
```
